src/BotPlutusInterface/CoinSelection.hs

```diff
@@ -0,0 +1,479 @@
+
+    -- the input vector for the current transaction, this can be a zero vector when there are no
+    -- inputs the transaction.
+    txInsVec <-
+      hoistEither $
+        if Value.isZero txInsValue
+          then Right $ zeroVec (length allAssetClasses)
+          else valueToVec allAssetClasses txInsValue
+
+    -- the output vector of the current transaction, this is all the values of TxOut combined.
+    outVec <- hoistEither $ valueToVec allAssetClasses outValue
+
+    -- all the remainingUtxos converted to the vectors.
+    remainingUtxosVec <- hoistEither $ mapM (valueToVec allAssetClasses . txOutValue . snd) remainingUtxos
+
+    -- we use the default search strategy to get indexes of optimal utxos, these indexes are for the
+    -- remainingUtxos, as we are sampling utxos from that set.
+    selectedUtxosIdxs <- newEitherT $ searchTxIns @w def (isSufficient outVec) outVec txInsVec remainingUtxosVec
+
+    lift $ printBpiLog @w (Debug [CoinSelectionLog]) $ "" <+> "Selected UTxOs Index: " <+> pretty selectedUtxosIdxs
+
+    let -- These are the selected utxos that we get using `selectedUtxosIdxs`.
+        selectedUtxos :: [(TxOutRef, TxOut)]
+        selectedUtxos = selectedUtxosIdxs ^.. folded . to (\idx -> remainingUtxos ^? ix idx) . folded
+
+        selectedVectors :: [Vector Integer]
+        selectedVectors = selectedUtxosIdxs ^.. folded . to (\idx -> remainingUtxosVec ^? ix idx) . folded
+
+    finalTxInputVector <- hoistEither $ foldM addVec txInsVec selectedVectors
+    unless (isSufficient outVec finalTxInputVector) $ throwError "Insufficient Funds"
+
+    selectedTxIns <- hoistEither $ mapM txOutToTxIn selectedUtxos
+
+    lift $ printBpiLog @w (Debug [CoinSelectionLog]) $ "Selected TxIns: " <+> pretty selectedTxIns
+
+    -- Now we add the selected utxos to originalTxIns present in the transaction previously.
+    return $ originalTxIns <> Set.fromList selectedTxIns
+  where
+    -- This represents the condition when we can stop searching for utxos.
+    -- First condition is that the input vector must not be zero vector, i.e.
+    -- There must be atleast some input to the transaction.
+    -- Second condition is that all the values of input vector must be greater than
+    -- or equal to the output vector.
+    isSufficient :: Vector Integer -> Vector Integer -> Bool
+    isSufficient outVec txInsVec =
+      Vec.all (== True) (Vec.zipWith (<=) outVec txInsVec)
+        && txInsVec /= zeroVec (length txInsVec)
+
+-- `searchTxIns` searches for optimal utxos for a transaction as input given
+-- current input vector, output vector and a list of all the remaining utxo vectors.
+searchTxIns ::
+  forall (w :: Type) (effs :: [Type -> Type]).
+  Member (PABEffect w) effs =>
+  SearchStrategy ->
+  (Vector Integer -> Bool) ->
+  Vector Integer ->
+  Vector Integer ->
+  [Vector Integer] ->
+  Eff effs (Either Text [Int])
+searchTxIns searchStrategy stopSearch outVec txInsVec utxosVec
+  | searchStrategy == Greedy =
```

Comment:
Use pattern-matching on the constructors directly instead of boolean guards -- the last guard is unnecessary since there are only two options and this can be statically guaranteed

Reply:
fixed